protocol.go

```diff
@@ -31,22 +31,42 @@ var (
 	ErrTokenValidationError = errors.New("token validation error")
 )
 
+// SymmetricKey Generic Symmetric Key
+type SymmetricKey interface {
+
+	// Encrypt encrypts a token with a symmetric key.
+	encrypt(payload interface{}, footer interface{}, unitTestNonce []byte) (string, error)
+
+	// Decrypt decrypts a token which was encrypted with a symmetric key.
+	decrypt(token string, payload interface{}, footer interface{}) error
+}
+
+// AsymmetricSecretKey Generic Asymmetric Private Key
+type AsymmetricSecretKey interface {
+
+	// Sign signs a token with the given private key
+	sign(payload interface{}, footer interface{}) (string, error)
+}
+
+// AsymmetricPublicKey Generic Asymmetric Public Key
+type AsymmetricPublicKey interface {
+
+	// Verify verifies a token against the given public key
+	verify(token string, value interface{}, footer interface{}) error
+}
+
 // Protocol defines the PASETO token protocol interface.
 type Protocol interface {
 
-	// Encrypt encrypts a token with a symmetric key. The key should be a byte
-	// slice of 32 bytes, regardless of whether PASETO v1 or v2 is being used.
-	Encrypt(key []byte, payload interface{}, footer interface{}) (string, error)
+	// Encrypt encrypts a token with a symmetric key
+	Encrypt(key SymmetricKey, payload interface{}, footer interface{}) (string, error)
```

Comment:
Something perhaps to consider here is how to handle the `implicit` param added the `encrypt`, `decrypt`, `sign`, and `verify` in v3 and v4. It probably isn't ideal to add an `implicit` arg to the existing methods in v1 and v2, as it may falsely give the impression that the `implicit` param is being validated (when it would be ignored for those versions). Perhaps a reasonable middle ground would be to add the param everywhere, but error if implicit is non-empty in v1 and v2. An alternative would be to specify separate methods altogether e.g. `EncryptImplicit(key SymmetricKey, payload interface{}, footer interface{}, implicit interface{}) (string, error)`, where these implicit methods always throw for v1 and v2.